src/main/java/ui/components/pickers/LabelPickerUILogic.java

```diff
@@ -59,58 +71,146 @@ private void populatePanes() {
         if (dialog != null) dialog.populatePanes(getExistingLabels(), getNewTopLabels(), bottomLabels, groups);
     }
 
-    public void toggleLabel(String name) {
-        addRemovePossibleLabel("");
-        preProcessAndUpdateTopLabels(name);
-        updateBottomLabels(""); // clears search query, removes faded-out overlay on bottom labels
+    public ArrayList<String> toggleLabel(String name) {
+        removePossibleLabel();
+        ArrayList<String> removedLabels = processAndUpdateTopLabels(name);
+        restoreBottomLabels(); // clears search query, removes faded-out overlay on bottom labels
         populatePanes();
+        return removedLabels;
     }
 
-    public void toggleSelectedLabel(String text) {
-        if (!bottomLabels.isEmpty() && !text.isEmpty() && hasHighlightedLabel()) {
-            toggleLabel(
-                    bottomLabels.stream().filter(PickerLabel::isHighlighted).findFirst().get().getActualName());
-        }
+    public ArrayList<String> toggleHighlightedLabel() {
+        return toggleLabel(getHighlightedLabelName().get().getActualName());
     }
 
-    @SuppressWarnings("PMD")
+    @SuppressWarnings("unused")
+    private void ______TEXT_FIELD______() {}
+
     public void processTextFieldChange(String text) {
-        String[] textArray = text.split(" ");
-        if (textArray.length > 0) {
-            String query = textArray[textArray.length - 1];
-            if (previousNumberOfActions != textArray.length || !query.equals(lastAction)) {
-                previousNumberOfActions = textArray.length;
-                lastAction = query;
-                boolean isBottomLabelsUpdated = false;
-
-                // group check
-                // TODO rewrite this to remove some nesting, and the PMD warning
-                if (TurboLabel.getDelimiter(query).isPresent()) {
-                    String delimiter = TurboLabel.getDelimiter(query).get();
-                    String[] queryArray = query.split(Pattern.quote(delimiter));
-                    if (queryArray.length == 1) {
-                        isBottomLabelsUpdated = true;
-                        updateBottomLabels(queryArray[0], "");
-                    } else if (queryArray.length == 2) {
-                        isBottomLabelsUpdated = true;
-                        updateBottomLabels(queryArray[0], queryArray[1]);
-                    }
-                }
+        if (isEmpty(text)) {
+            removePossibleLabel();
+            restoreBottomLabels();
+        } else {
+            processChangeAndUpdateVariables(text);
+        }
+        populatePanes();
+    }
 
-                if (!isBottomLabelsUpdated) {
-                    updateBottomLabels(query);
-                }
+    private void processChangeAndUpdateVariables(String text) {
+        if (isLongerTextFieldThanPrevious(text)) {
+            addCharHandler(text);
```

Comment:
Did you mean `handleCharAddition`?. The current name sounds like you are adding a char handler.